### ISSUE TITLE
Attachmentfiles on devices are rw - changes to attachements are not signaled to the user by DeltaChat

### ISSUE THREAD

**Issue by menthenchocolat** [deltachat/deltachat-desktop#4352]
- Operating System (Linux/Mac/Windows/iOS/Android): Linux
- Delta Chat Version: 1.48

- Expected behavior:
If I click on an attachment (e.g. a LibreOffice document) sent to me via DeltaChat, DeltaChat either prompts me to save this to some folder on my device, or the attachment is opened directly in e.g. LibreOffice, but LibreOffice tells me, that the file is read-only and needs to be saved somewhere else before I can edit it.

- Actual behavior:
I click on an attachment in DeltaChat, attachment is opened directly in e.g. LibreOffice and I can edit & save directly to the attachment file in the DeltaChat storage location. Thus data between me and sender becomes inconsistent. DeltaChat does not inform me about the edited attachment.
Still worse is what happened to a friend of mine: He edited the document, saved it. The save was persistent, but only for a day or so. Then the changes in the document got reverted (by DeltaChat?) to the original version.

- Steps to reproduce the problem:
Open attachment in DeltaChat, edit, save (not "save as ..." but "save"). Then: reopen in DeltaChat. Edit is persistent, meaning attachment is not anymore in the state it was sent. DeltaChat does not inform user about this.
I was unable to reproduce the worse version on my machine. I believe what happened on my friends machine was due to him using a mail-server setup different from mine (him using the default DeltaChat-Server, me using my own mail provider).

- Screenshots: -
- Logs: -

- Comment: I know that one should save attachments to a different location before editing them, but there are many NOOBS out there and DeltaChats behaviour (see above) will get a lot of them in trouble. They should be able to use DeltaChat safely, I believe.

**Comment by nicodh** [comment]
> When the user clicks on an attachment to open it:
> 
> * Copy the attachment file to a temporary directory. The MessageObject returned by `get_message(…)` includes a `file_name`, which should be used as the name.
> * Open the new, temporary file instead of the original.
> 
> Make sure that the temporary directory is cleaned afterwards.

I'm not sure if that would be expected behaviour. So if I open a file, edit it and close it I would expect my changes to be persisted. If the temporary directory is cleaned after that the changes will be lost again.

I would suggest to show a dialog:
Do you want to save the file or open it "read only"?

Then everything is transparent for the user.

@Hocuri @r10s

**Comment by r10s** [comment]
hm, in general, we try to avoid such upfront questions as much as possible.

also, the vast majority of cases, user would will  open the file read-only, so that seems to be a better default.

so, i would make the copy read-only and call it a day. most external apps will show the state and _then_ user can use sth. as "Save As"

**Comment by nicodh** [comment]
> Instead of copying to temporary directory (e.g. `/tmp`), maybe make a copy to Downloads folder?

The only downside is that these files will stay there, which also may be unexpected if a user just reads them. Let's say someone sends me a PDF. I click on it read it, close the reader and delete the message (or even the whole account): I would not expect the PDF still to be in my DownloadFolder. (like I would in the Browser)

Another problem: if we open a file let's say in a PDF reader, keep it open and then open another version sent by someone with the same name we have a problem, when copying it to the same destination.

In tmp dir my idea was to have an additional folder with the random filename as it is persisted in blobdir and copy the file into that. But in Download that would look weird.

And it is anyway not clear when to delete the file again (in tmp dir). We don't have a routine for that afaik.

So finally - make the files in blobdir readonly seems the easiest solution. But that might prevent people from deleting it, depending on the OS

**Comment by r10s** [comment]
i am also not sure about saving everything to the "Download" folder, that may be quite unexpected, indeed. "Viewing" an already local file is usually not regarded as "Downloading".

> make the files in blobdir readonly seems the easiest solution

iirc (cc @Hocuri) this was not the way we want to go when opening eg. a PDF. also as the name is messed up then, and the user cannot use "Save As" within the app.

so, imu of previous discussions, we create a read-only-copy in tmp-dir using the correct file name and pass that to system's "open" command

**Comment by r10s** [comment]
still not 100% sure, however, we can also saving to "Downloads" a try, esp. if it is simpler at the end. maybe it is fine and no one cares about the files in "Download". i would not die on that hill :)

**Comment by Hocuri** [comment]
I agree with @r10s and @nicodh that usability-wise, saving to a temporary directory is nicest, but saving to Downloads is also fine.

> And it is anyway not clear when to delete the file again (in tmp dir). We don't have a routine for that afaik.

We can just delete whatever file is in this temp directory when DC is launched and closed (or only when it's closed), this way it will be deleted soon enough. It doesn't need to be in /tmp, it could be a folder in the DeltaChat folder, too - whatever is easiest to implement.

BTW, probably the temp dir doesn't even need to have a random name - you can just create one temp dir for this, and if there already is a file in there, delete it. Usually the user will only open one file at a time, and when the user does open another file from DC, almost all external programs will react fine & just continue showing the file even after it was deleted.

> make the files in blobdir readonly seems the easiest solution

Note that DC Desktop must not simply mark files in the blobdir as readonly, because core will fail to delete it, and core will fail to receive another file with the same content. If we actually go for the read-only-solution, then reverting https://github.com/deltachat/deltachat-core-rust/commit/9bc4071463bd65ce3a806c192e222cabe091017e might be the way to go.

_But_ I wouldn't give up on the correct filename that quickly. Also, a file named `archive.tar.xz` will become e.g. `9e7f409ac5c92b942cc4f31cee2770a.xz`, i.e. the `.tar` will be lost. Is there anything speaking against copying to a (non-random) temp folder, if there already is a file in there then deleting it, and clearing the temp folder on startup and closing?

**Comment by nicodh** [comment]
What about a tmp_files dir in the account folder, which is cleared on next start?
 - flatpak should be fine with that
 - when an account is deleted the tmp files will be too 
 - it is not as unexpectd to have files there which belong to the app and the account, compared to a Download dir

**Comment by Hocuri** [comment]
> I also looked what Signal Desktop does: it simply shows a save dialog offering to save PDFs (and office docs). Maybe it's an ok thing to do.

I'm using Signal Desktop and I hate this because it feels way too cubersome to open a PDF, but maybe it's just me and it's OK for most people.

Anyway, I think we should go with what @nicodh described in his latest post.

**Comment by menthenchocolat** [comment]
I use thunderbird for a couple of years now, and I never ran into that kind of problem with email-attachments there. So I just checked what Thunderbird does, e.g.: "What happens, if I click on a LibreOffice document attached to an email?"

Found the following (on Linux):
The email-attachment is copied with its original name and read-only to a folder called /tmp/pid-nnnnn/ (where nnnnn is the number of the process running thunderbird). It is then opened there.

Benefit (in my view): File is read-only, so my dad, when he want's to change something in the document, automatically has to save it to somewhere else before closing LibreOffice.

If it was not read-only in the '/tmp/pid-nnnnn'-dir, we would run into another problem: My dad would save his changes to the file there. Those changes would vanish with the next flush of the '/tmp'-dir.

To me, this behavior of thunderbird on Linux seems reasonable and NOOB-safe.

Does anybody know about the behavior of thunderbird on other OSes than Linux?
